Sources/MapboxCoreNavigation/CoreNavigationNavigator.swift

```diff
@@ -41,19 +41,24 @@ class Navigator {
         }
     }
     
-    let historyRecorder: HistoryRecorderHandle
+    private(set) var historyRecorder: HistoryRecorderHandle!
     
-    let navigator: MapboxNavigationNative.Navigator
+    private(set) var navigator: MapboxNavigationNative.Navigator!
```

Comment:
Force unwrapping is very dangerous practice is should be avoided if possible.

I join @bamx23 and suggest to extract these variables into class. Then you can have static method which will create these dependencies and then use this static factory method in `init` and `restart`. 

Here is a variant with tuples just to make it clearer what I suggest:

```swift
    static func create(version: String?) -> (tileVersionState: TileVersionState, tileStore: TileStore, historyRecorder: HistoryRecorderHandle, cacheHandle: CacheHandle, roadGraph: RoadGraph, navigator: MapboxNavigationNative.Navigator) {
        let tileVersionState = TileVersionState.nominal

        let settingsProfile = SettingsProfile(application: ProfileApplication.kMobile,
                                              platform: ProfilePlatform.KIOS)

        let endpointConfig = TileEndpointConfiguration(credentials:Navigator.credentials ?? Directions.shared.credentials,
                                                       tilesVersion: version ?? Self.tilesVersion,
                                                       minimumDaysToPersistVersion: nil,
                                                       isFallback: version != nil)

        let tileStorePath = Self.tilesURL?.path ?? ""
        let tileStore = TileStore.getInstanceForPath(tileStorePath)
        let tilesConfig = TilesConfig(tilesPath: tileStorePath,
                                      tileStore: tileStore,
                                      inMemoryTileCache: nil,
                                      onDiskTileCache: nil,
                                      mapMatchingSpatialCache: nil,
                                      threadsCount: nil,
                                      endpointConfig: endpointConfig)

        let historyAutorecordingConfig = [
            "features": [
                "historyAutorecording": true
            ]
        ]

        var customConfig = ""
        if let jsonDataConfig = try? JSONSerialization.data(withJSONObject: historyAutorecordingConfig, options: []),
           let encodedConfig = String(data: jsonDataConfig, encoding: .utf8) {
            customConfig = encodedConfig
        }

        let configFactory = ConfigFactory.build(for: settingsProfile,
                                                config: NavigatorConfig(),
                                                customConfig: customConfig)

        let historyRecorder = HistoryRecorderHandle.build(forHistoryFile: Navigator.historyDirectoryURL?.path ?? "", config: configFactory)

        let runloopExecutor = RunLoopExecutorFactory.build()
        let cacheHandle = CacheFactory.build(for: tilesConfig,
                                         config: configFactory,
                                         runLoop: runloopExecutor,
                                         historyRecorder: historyRecorder)

        let roadGraph = RoadGraph(MapboxNavigationNative.GraphAccessor(cache: cacheHandle))

        let navigator = MapboxNavigationNative.Navigator(config: configFactory,
                                                     runLoopExecutor: runloopExecutor,
                                                     cache: cacheHandle,
                                                     historyRecorder: historyRecorder)

        return (tileVersionState: tileVersionState, tileStore: tileStore, historyRecorder: historyRecorder, cacheHandle: cacheHandle, roadGraph: roadGraph, navigator: navigator)
    }
```

Reply:
Force unwrapping is used to allow initializing properties using a method call inside `init()`.
I've been thinking in a similar way in [subsequent PR](https://github.com/mapbox/mapbox-navigation-ios/pull/3015/files#diff-fb53bdfdf70730b6dbbb59d044c859accd4e7709e8857bab1462d493298e265bR76), where this approach is put to real use-case.
I didn't consider introduce a dedicated type instead of a tuple to wrap all the properties because I couldn't come up with a decent naming for the context of such type.

Reply:
I want to encourage you to remove force unwraps and use the work from this [next PR](https://github.com/mapbox/mapbox-navigation-ios/pull/3015/files#diff-fb53bdfdf70730b6dbbb59d044c859accd4e7709e8857bab1462d493298e265bR76). 

Personally, I ok with using tuples. 

I've seen how force unwraps results in crashes in production, it is effortless to break during modifications, unreliable, and the compiler doesn't help you. In other words, they are very hard to maintain due to relaxed support from the Swift compiler.

Reply:
Introduced `NativeHandlersFactory` for that functionality.